extensions/amp-analytics/0.1/iframe-transport-message-queue.js

```diff
@@ -0,0 +1,126 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+import {
+  IFRAME_TRANSPORT_EVENTS_TYPE,
+} from '../../../src/3p-analytics-common';
+import {SubscriptionApi} from '../../../src/iframe-helper';
+
+/** @private @const {string} */
+const TAG_ = 'amp-analytics.IframeTransportMessageQueue';
+
+/** @private @const {number} */
+const MAX_QUEUE_SIZE_ = 100;
+
+/**
+ * @visibleForTesting
+ */
+export class IframeTransportMessageQueue {
+  /**
+   * Constructor
+   * @param {!Window} win The window element
+   * @param {!HTMLIFrameElement} frame The cross-domain iframe to send
+   * messages to
+   */
+  constructor(win, frame) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!HTMLIFrameElement} */
+    this.frame_ = frame;
+
+    /** @private {boolean} */
+    this.isReady_ = false;
+
+    /**
+     * @private {!Array<!../../../src/3p-analytics-common.IframeTransportEvent>}
+     */
+    this.pendingEvents_ = [];
+
+    /** @private {string} */
+    this.messageType_ = IFRAME_TRANSPORT_EVENTS_TYPE;
```

Comment:
Ah, I see what you mean. The reason this is sent from iframe to host, is that the iframe is saying "I want to subscribe to this". Per SubscriptionAPI, they therefore must be the same. See listenForListener at https://github.com/ampproject/amphtml/blob/master/src/iframe-helper.js#L198